src/ot/normalize.rs

```diff
@@ -0,0 +1,500 @@
+//
+// The goals are:
+//
+//   - Try to render all canonically equivalent strings similarly.  To really
+//     achieve this we have to always do the full decomposition and then
+//     selectively recompose from there.  It's kinda too expensive though, so
+//     we skip some cases.  For example, if composed is desired, we simply
+//     don't touch 1-character clusters that are supported by the font, even
+//     though their NFC may be different.
+//
+//   - When a font has a precomposed character for a sequence but the 'ccmp'
+//     feature in the font is not adequate, use the precomposed character
+//     which typically has better mark positioning.
+//
+//   - When a font does not support a combining mark, but supports it precomposed
+//     with previous base, use that.  This needs the itemizer to have this
+//     knowledge too.  We need to provide assistance to the itemizer.
+//
+//   - When a font does not support a character but supports its canonical
+//     decomposition, well, use the decomposition.
+//
+//   - The complex shapers can customize the compose and decompose functions to
+//     offload some of their requirements to the normalizer.  For example, the
+//     Indic shaper may want to disallow recomposing of two matras.
+
+pub struct ShapeNormalizeContext {
+    pub plan: ot::ShapePlan,
+    pub(crate) buffer: &'static mut Buffer,
+    pub font: &'static Font<'static>,
+    pub decompose: ffi::rb_ot_decompose_func_t,
+    pub compose: ffi::rb_ot_compose_func_t,
+}
+
+impl ShapeNormalizeContext {
+    #[inline]
+    pub fn from_ptr(ctx: *const ffi::rb_ot_shape_normalize_context_t) -> &'static Self {
+        unsafe { &*(ctx as *const Self) }
+    }
+
+    #[inline]
+    pub fn as_ptr(&self) -> *const ffi::rb_ot_shape_normalize_context_t {
+        self as *const _ as *const ffi::rb_ot_shape_normalize_context_t
+    }
+
+    #[inline]
+    pub fn decompose(&self, ab: u32) -> Option<(u32, u32)> {
+        let mut a = 0;
+        let mut b = 0;
+
+        unsafe {
+            if (self.decompose)(self.as_ptr(), ab, &mut a, &mut b) != 0 {
+                return Some((a, b));
+            }
+        }
+
+        None
+    }
+
+    #[inline]
+    pub fn compose(&self, a: u32, b: u32) -> Option<u32> {
+        let mut ab = 0;
+
+        unsafe {
+            if (self.compose)(self.as_ptr(), a, b, &mut ab) != 0 {
+                return Some(ab);
+            }
+        }
+
+        None
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[allow(dead_code)]
+pub enum ShapeNormalizationMode {
+    None = 0,
+    Decomposed,
+    /// Never composes base-to-base.
+    ComposedDiacritics,
+    /// Always fully decomposes and then recompose back.
+    ComposedDiacriticsNoShortCircuit,
+    Auto,
+}
+
+impl Default for ShapeNormalizationMode {
+    fn default() -> Self {
+        Self::Auto
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn _rb_ot_shape_normalize(
```

Comment:
Could you wrap extern functions?

So it would be like this:

```rust
#[no_mangle]
pub extern "C" fn _rb_ot_shape_normalize(
    plan: *const ffi::rb_ot_shape_plan_t,
    buffer: *mut ffi::rb_buffer_t,
    font: *mut ffi::rb_font_t,
) {
    let plan = ot::ShapePlan::from_ptr(plan);
    let buffer = Buffer::from_ptr_mut(buffer);
    let font = Font::from_ptr(font);
    normalize(plan, buffer, font);
}

fn normalize(...) { /* actual code */ }
```

Reply:
Sure, I also did it for the fallback shaper functions.